pkg/controller/control.go

```diff
@@ -1285,3 +1343,79 @@ func (c *Controller) GetLatestMetics() *enginerpc.Metrics {
 func getAverageLatency(totalLatency, iops uint64) uint64 {
 	return totalLatency / iops
 }
+
+// tryFreeze attempts to bind mount an existing mount point to freezePoint, then freeze the filesystem from
+// freezePoint. tryFreeze returns booleans mounted and frozen so the caller can attempt to clean up depending on its
+// progress. tryFreeze does not return an error if there are no eligible mount points to freeze.
+func tryFreeze(devicePath string, freezePoint string, mounter mount.Interface, exec lhexec.ExecuteInterface,
+	log logrus.FieldLogger) (mounted, frozen bool, err error) {
+	// We do not need to switch to the host mount namespace to get mount points here. Usually, longhorn-engine runs in a
+	// container that has / bind mounted to /host with at least HostToContainer (rslave) propagation.
+	// - If it does not, we likely can't do a namespace swap anyway, since we don't have access to /host/proc.
+	// - If it does, we just need to know where in the container we can access the mount points to create a bind mount.
+	sourcePoints, err := mounter.List()
+	if err != nil {
+		log.WithError(err).Warn("Failed to list mount points while deciding whether or not to freeze")
+	}
+
+	for _, sourcePoint := range sourcePoints {
+		if sourcePoint.Device != devicePath {
+			continue // We are iterating through the list of all mounted filesystems.
+		}
+		log.Debugf("Mounting filesystem to %v", freezePoint)
+		if err = attemptBindMountFilesystem(sourcePoint.Path, freezePoint, mounter); err != nil {
+			// There's no particular reason to think we will be successful with a different sourcePoint.
+			err = errors.Wrapf(err, "failed to bind mount filesystem from %v to %v", sourcePoint, freezePoint)
+			return
+		}
+		mounted = true
+
+		// We must verify it is still save to freeze the filesystem. If the source mount was unmounted by someone else
+		// before we bind mounted, the bind mount could refer to the root filesystem.
+		var device string
+		device, _, err = mount.GetDeviceNameFromMount(mounter, freezePoint)
+		if err != nil || device != devicePath {
+			// This would likely only happen if an upper layer unmounted the volume. If that is the case, we are
+			// probably being torn down. Trying other sourcePoints will likely lead to a pointless race.
+			err = errors.Wrapf(err, "cannot verify mount point %v refers to volume", freezePoint)
+			return
+		}
```

Comment:
Should we directly find the filesystem on the root partition of the `devicePath` and mount it to `freezePoint` so we don't have to find and bind mount the existing mount points? And if there is no filesystem on the root partition, we can do no-op and fallback to sync.

By doing this, we don't have to worry that the upper layer is unmounting while we are doing bind mount and attempt to freeze

Reply:
I thought about this, and I don't feel too strongly either way. If a different implementation did as you suggested, it would probably be fine to me. Here are a couple reasons the bind mount approach might be better.

- If the file system is not mounted, freezing it is unnecessary. The bind mount approach only applies to a mounted file system.
- I'm not 100% sure of the behavior when directly mounting a file system that is already mounted. I tried it for ext4 and XFS volumes in my Rocky environment and it seemed to have the same effect as a bind mount. However, CSI drivers (including ours) always bind mount the staging path to the target path instead of doing a direct mount to the target path. It seems safer to me to follow suit.